compiler/AST/AstToText.cpp

```diff
@@ -0,0 +1,1318 @@
+/*
+ * Copyright 2004-2015 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "AstToText.h"
+
+#include "expr.h"
+#include "stmt.h"
+#include "symbol.h"
+
+AstToText::AstToText()
+{
+
+}
+
+AstToText::~AstToText()
+{
+
+}
+
+const std::string& AstToText::text() const
+{
+  return mText;
+}
+
+void AstToText::appendNameAndFormals(FnSymbol* fn)
+{
+  if (fn->instantiatedFrom != NULL && developer == false)
+  {
+    appendNameAndFormals(fn->instantiatedFrom);
+  }
+  else
+  {
+    appendName(fn);
+    appendFormals(fn);
+  }
+}
+
+/************************************ | *************************************
+*                                                                           *
+* Append a normalized version of the function's name to the buffer.         *
+*                                                                           *
+************************************* | ************************************/
+
+void AstToText::appendName(FnSymbol* fn)
+{
+  if (developer == true)
+    mText += fn->name;
+
+  else if (fn->hasFlag(FLAG_MODULE_INIT))
+  {
+    INT_ASSERT(strncmp(fn->name, "chpl__init_",      11) == 0);
+
+    mText += "top-level module statements for ";
+    mText += (fn->name + 11);
```

Comment:
These constants are a little magic right now.  While I wouldn't recommend determining them from the expected prefix (string ops take so long), it might be better to store them for repeated use with a note as to what they represent.